src/Orleans/Providers/DefaultServiceProvider.cs

```diff
@@ -8,8 +10,47 @@ namespace Orleans.Providers
     /// </summary>
     public class DefaultServiceProvider : IServiceProvider
     {
+        private readonly IDictionary<Type, object> singletonsByType;
+        private readonly IDictionary<Type, object> singletonsByInterface;
+
+        public DefaultServiceProvider()
+            : this(new Dictionary<Type,object>() )
+        {
+        }
+
+        public DefaultServiceProvider(IDictionary<Type, object> singletons)
+        {
+            if (singletons == null)
+            {
+                throw new ArgumentNullException("singletons");
+            }
+
+            // If type is interface, add to list of singletons obtainable via interface
+            singletonsByInterface = singletons.Where(kvp => kvp.Key.IsInterface)
+                .ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+
+            // If type is not interface, add to list of singletons accessible by type
+            singletonsByType = singletons.Where(kvp => !kvp.Key.IsInterface)
+                .ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+
+            // Regardless of type mapping provided, always map singleton to it's actual type
+            foreach (object singleton in singletons.Values)
+            {
+                singletonsByType[singleton.GetType()] = singleton;
```

Comment:
this is probably different to most container implementations, they don't just register the concrete object for you if you don't explicitly do it